pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/MultipleRollupPinotSegmentRecordReader.java

```diff
@@ -0,0 +1,233 @@
+    return next(new GenericRow());
+  }
+
+  @Override
+  public GenericRow next(GenericRow reuse) throws IOException {
+    // Copy the current row to reuse, which will be the final result
+    RecordReaderUtils.copyRow(_currentRow, reuse);
+
+    while (true) {
+      // If there is no more row left, mark the current row to be null and break the loop
+      if (!_multipleSegmentRecordReader.hasNext()) {
+        _currentRow = null;
+        break;
+      }
+
+      // Read the current row
+      _currentRow = convertTimeGranularity(_multipleSegmentRecordReader.next(_currentRow));
+
+      // If the dimension and the time column value is not the same, break the loop to return the result
+      if (!haveSameDimensionAndTimeColumn(reuse, _currentRow)) {
+        break;
+      } else {
+        // Since the dimension and time columns are the same, metric values need to be aggregated
+        for (MetricFieldSpec metric : _schema.getMetricFieldSpecs()) {
+          String metricName = metric.getName();
+          ValueAggregator aggregator = (_aggregatorMap == null) ? DEFAULT_AGGREGATOR_FUNCTION
+              : _aggregatorMap.getOrDefault(metricName, DEFAULT_AGGREGATOR_FUNCTION);
+          Object aggregatedResult =
+              aggregator.aggregate(reuse.getValue(metricName), _currentRow.getValue(metricName), metric);
+          reuse.putField(metricName, aggregatedResult);
+        }
+      }
+    }
+    return reuse;
+  }
+
+  @Override
+  public void rewind() throws IOException {
+    _multipleSegmentRecordReader.rewind();
+    if (_multipleSegmentRecordReader.hasNext()) {
+      _currentRow = convertTimeGranularity(_multipleSegmentRecordReader.next());
+    }
+  }
+
+  @Override
+  public Schema getSchema() {
+    return _schema;
+  }
+
+  @Override
+  public void close() throws IOException {
+    _multipleSegmentRecordReader.close();
+  }
+
+  /**
+   * Helper function to convert the time column value based on the input date time transformer
+   */
+  private GenericRow convertTimeGranularity(GenericRow row) {
+    String timeColumn = _schema.getTimeColumnName();
+    if (_dateTimeTransformer == null || timeColumn == null) {
+      return row;
+    }
+    FieldSpec.DataType dataType = _schema.getTimeFieldSpec().getDataType();
+    switch (dataType) {
+      case INT:
+        _timeValueInput[0] = (Integer) row.getValue(_schema.getTimeColumnName());
+        _dateTimeTransformer.transform(_timeValueInput, _timeValueOutput, 1);
+        row.putField(timeColumn, (int) _timeValueOutput[0]);
+        break;
+      case LONG:
+        _timeValueInput[0] = (Long) row.getValue(_schema.getTimeColumnName());
+        _dateTimeTransformer.transform(_timeValueInput, _timeValueOutput, 1);
+        row.putField(timeColumn, _timeValueOutput[0]);
+        break;
+      default:
+        throw new IllegalStateException(
+            "Time column granularity change is only supported for a time column with int or long data type");
```

Comment:
Why are we skipping out String? For simple date format, the time column might be in string right?

Reply:
I see that the config itself has only EpochToEpochTransformer. Why are we sticking to just epoch input format?